DSA code

chrehall68 opened the pull request:
This PR contains every python example. They were copied over from the google docs.

chrehall68 committed:
transfer chapter 2 and 3 code

chrehall68 committed:
copy chapter 1 code

lint-action committed:
Fix code style issues with Black

chrehall68 committed:
fix flake errors

chrehall68 committed:
Merge branch 'dsa' of https://github.com/code4tomorrow/python into dsa

chrehall68 committed:
fix flake errors in linked_list.py

chrehall68 commented on dsa/chapter1/recursion.py:
```diff
@@ -0,0 +1,24 @@
+
+
+# recursive function
+def offset(x):
```
I copied this from the google doc. However, this isn't a recursive function. Should we remove this?

Chrovo commented on dsa/chapter2/bst.py:
```diff
@@ -0,0 +1,141 @@
+class BinaryTree:
+    class Node:
```
I'd put the `Node` class outside of the `BinaryTree` class. Students might get lost with indentation while reading the code and mix up what code belongs to which class. It also looks better visually in my opinion.

chrehall68 committed:
add in the other partitions

hpanchal092 committed:
take Node class outside of BinaryTree class
 - also fixed recursive comment

hpanchal092 committed:
updated recursion code to match changes in notion
 - changed recursion lesson in chapter 1 because it was bad
 - new code reflects the changes in notion

chrehall68 committed:
an example a_star, implemented where you can move on diagonals and in cardinal directions

chrehall68 committed:
upload graph code

chrehall68 committed:
fix flake error

hpanchal092 reviewed (approved):
LGTM 👍

chrehall68 committed:
remove extra +1

shiventi reviewed (commented):
LGTM!!!
